working/bazel/module-structure.md

```diff
@@ -0,0 +1,226 @@
+    name = "flags",
+    srcs = ["lib/flags.dart"],
+)
+
+dart_test(
+    name = "flags_test",
+    srcs = ["test/flags_test.dart"],
+    deps = [
+        ":flags",
+        "//third_party/dart/test",
+    ],
+)
+```
+
+Here we have _3_ targets:
+* `app`, which potentially is the entrypoint to our package/application.
+* `flags`, which contains some common code for setting/getting flags.
+* `flags_test`, which tests that `flags` is working-as-intended.
+
+This concept already is very different than a `pub` package, where all of the
+files in `lib/` are sort of a single "implicit" target. In fact, a common issue
+externally is that `pubspec.yaml` (sort of similar to `BUILD`) is not granular
+enough, leading to the creation of "micro packages" that have a single file or
+capability.
+
+### Common Patterns
+
+Based on the above, teams tend to structure their projects _hierarchically_,
+with more specific code living deeper in a sub-package. For example, imagine
+the `ninjacat` project after a few more weeks:
+
+```
+> ninjacat/
+  > app/
+    > views/
+      > checkout/
+        > lib/
+        > test/
+        > BUILD
+      > home/
+        > lib/
+        > test/
+        > BUILD
+  > common/
+    > widgets/
+      > login/
+        > lib/
+        > test/
+        > testing/
+          > lib/
+          > BUILD
+        > BUILD
+```
+
+Already we have the following "packages":
+
+* `package:ninjacat.app.views.checkout`
+* `package:ninjacat.app.views.home`
+* `package:ninjacat.common.widgets.login`
+* `package:ninjacat.common.widgets.login.testing`
+
+It's common to use the [`visibility`][5] property to setup the concept of
+"friend" packages, i.e. packages that are only accessible to _specific_ other
+packages. In the following code, we ensure that the `login` "package" is only
+accessible to packages under `app/`:
+
+```
+# //ninjacat/common/widgets/login/BUILD
+
+dart_library(
+    name = "login",
+    srcs = glob(["lib/**.dart"]),
+    visibility = [
+        "//ninajacat/app:__subpackages__",
+    ],
+)
+```
+
+[5]: https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes
+
+Another common pattern is the concept of `testonly`, and `testing` packages:
+
+```
+# //ninjacat/common/widgets/login/BUILD
+
+dart_test(
+    name = "login_test",
+    srcs = ["test/login_test.dart"],
+    deps = [
+        "//ninjacat/common/widgets/login/testing",
+        "//third_party/dart/test",
+    ],
+)
+```
+
+```
+# //ninjacat/common/widgets/login/testing
+
+dart_library(
+    name = "testing",
+    srcs = glob(["lib/**.dart"]),
+    testonly = 1,
+    deps = [
+        "//third_party/dart/pageloader",
+    ],
+)
+```
+
+In this case, `package:ninjacat.common.widgets.login.testing` exposes a set of
+test-only libraries that can be used, only in _test_ targets. This pattern isn't
+replicable externally at all, but is _very_ common internally to avoid bringing
+in test utilities and code into production applications.
```

Comment:
Interesting insight - I could imagine a `dev_only` flag which could be set which only allows a package to be used as a dev dependency or as a direct dependency of other `dev_only` packages - this would be non-breaking for pub (but a breaking change per package when they opt in).

Reply:
Ack